compiler/aarch64/codegen/OMRCodeGenerator.cpp

```diff
@@ -19,52 +19,313 @@
+      if (branch != NULL)
+         {
+         if (abs(branch->getEstimatedBinaryLocation() - branch->getLabelSymbol()->getEstimatedCodeLocation()) > 512 * 1024)
+            {
+            candidateBranches.add(branch);
+            }
+         }
+      cursorInstruction = cursorInstruction->getNext();
+      }
+
+   // The following heuristic algorithm penalizes backward branches in
+   // estimation, since it should be rare that a backward branch needs
+   // a far relocation.
+
+   for (int32_t i = candidateBranches.lastIndex(); i >= 0; i--)
+      {
+      int32_t myLocation = candidateBranches[i]->getEstimatedBinaryLocation();
+      int32_t targetLocation = candidateBranches[i]->getLabelSymbol()->getEstimatedCodeLocation();
+      int32_t j;
+
+      // aarch64 B.cond instruction can reach +/-1MB address range
+
+      if (targetLocation >= myLocation)
+         { // forward branch
+         for (j = i + 1;
+              j < candidateBranches.size() && targetLocation > candidateBranches[j]->getEstimatedBinaryLocation();
+              j++)
+            ;
+
+         if ((targetLocation-myLocation + (j-i-1)*4) >= 1024 * 1024)
+            {
+            candidateBranches[i]->setFarRelocation(true);
+            }
+         else
+            {
+            candidateBranches.remove(i);
+            }
+         }
+      else
+         { // backward branch
+         for (j = i - 1;
+              j >= 0 && targetLocation <= candidateBranches[j]->getEstimatedBinaryLocation();
+              j--)
+            ;
+
+         if ((myLocation-targetLocation + (i-j-1)*4) > 1024 * 1024)
+            {
+            candidateBranches[i]->setFarRelocation(true);
+            }
+         else
+            {
+            candidateBranches.remove(i);
+            }
+         }
+      }
+
+   return estimate + 4*candidateBranches.size();
    }
 
 void
 OMR::ARM64::CodeGenerator::doBinaryEncoding()
    {
-   TR_ASSERT(false, "Not implemented yet.");
+   TR::Compilation *comp = self()->comp();
+   int32_t estimate = 0;
+   TR::Instruction *cursorInstruction = self()->getFirstInstruction();
+
+   self()->getLinkage()->createPrologue(cursorInstruction);
+
+   bool skipOneReturn = false;
+   while (cursorInstruction)
+      {
+      if (cursorInstruction->getOpCodeValue() == TR::InstOpCode::ret)
+         {
+         if (skipOneReturn == false)
+            {
+            TR::Instruction *temp = cursorInstruction->getPrev();
+            self()->getLinkage()->createEpilogue(temp);
+            cursorInstruction = temp->getNext();
+            skipOneReturn = true;
+            }
+         else
+            {
+            skipOneReturn = false;
+            }
+         }
+      estimate = cursorInstruction->estimateBinaryLength(estimate);
+      cursorInstruction = cursorInstruction->getNext();
+      }
+
+   estimate = self()->setEstimatedLocationsForSnippetLabels(estimate);
+   if (estimate > 512 * 1024)
```

Comment:
Same comment as above for the materialization of the 512 constant.

Reply:
This should have been 1MB (1024 * 1024).  The p and arm32 versions use 32768 at this condition.